latte/latte.TESTGEN.ps1

```diff
@@ -77,99 +73,117 @@ function test_send {
         # - Measures over input samples
-        banner -Msg "Iteration Tests: [$soc] operations per bounded iterations"
-        [int] $tmp  = 50000
-        [int] $iter = 10000 # When "-i" is used, latte limits iterations to 1-10000 range.
-        foreach ($snd in $snds) {
-            for ($i=0; $i -lt $g_iters; $i++) {
-                [int] $portstart = $tmp + ($i * $g_iters)
-
-                #optimized
-                test_send -Iter "-i $iter" -Port $portstart -Type "-$soc" -Snd $snd -Options "-group 0 -rio -riopoll 100000000000" -OutDir $dirOptimized -Fname "$soc.i$iter.$snd.OPT.iter$i"
-                test_recv
-
-                # Default
-                if ($g_detail) {
-                    test_send -Iter "-i $iter" -Port $portstart -Type "-$soc" -Snd $snd -OutDir $dirDefault -Fname "$soc.i$iter.$snd.iter$i"
-                    test_recv
-                }
-            }
+        if ($g_Config.PingIterations -gt 0) {
+            banner -Msg "Iteration Tests: [$Protocol] operations per bounded iterations"
+            test_protocol -Iter "-i $($g_Config.PingIterations)" -Protocol $Protocol -OutDirOpt $dirOptimized -OutDirDefault $dirDefault -Fname "$Protocol.i$($g_Config.PingIterations)"
         }
-
         # Transactions per 10s
         # - Measures operations per bounded time.
-        banner -Msg "Time Tests: [$soc] operations per bounded time"
-        [int] $tmp = 50000
-        [int] $sec = 30
-        foreach ($snd in $snds) {
-            for ($i=0; $i -lt $g_iters; $i++) {
-                [int] $portstart = $tmp + ($i * $g_iters)
-
-                # Optimized
-                test_send -Iter "-t $sec" -Port $portstart -Type "-$soc" -Snd $snd -Options "-group 0 -rio -riopoll 100000000000" -OutDir $dirOptimized -Fname "$soc.t$sec.$snd.OPT.iter$i" -NoDumpParam $true
-                test_recv
-                
-                # Default
-                if ($g_detail) {
-                    test_send -Iter "-t $sec" -Port $portstart -Type "-$soc" -Snd $snd -OutDir $dirDefault -Fname "$soc.t$sec.$snd.iter$i" -NoDumpParam $true
-                    test_recv
-                }
-            }
+        if ($g_Config.Time -gt 0) {
+            banner -Msg "Time Tests: [$Protocol] operations per bounded time"
+            test_protocol -Iter "-t $($g_Config.Time)" -Protocol $Protocol -OutDirOpt $dirOptimized -OutDirDefault $dirDefault -Fname "$Protocol.t$($g_Config.Time)" -NoDumpParam $true
         }
     }
 } # test_latte_generate()
 
+function validate_config {
+    $isValid = $true
+    $int_vars = @('Iterations', 'StartPort', 'Time', 'PingIterations')
+    foreach ($var in $int_vars) {
+        if (($null -eq $g_Config.($var)) -or ($g_Config.($var) -lt 0)) {
+            Write-Host "$var is required and must be greater than or equal to 0"
+            $isValid = $false
+        }
+    }
+    $valid_protocols = @('tcp', 'udp', 'raw')
+    foreach ($proto in $g_Config.Protocol) {
```

Comment:
thinking aloud here.. if the config file has no Protocol field then it wont enter this for loop and will pass the validate checks.. should we also add a check for if $g_Config.Protocol ?

Reply:
same applies to others